controllers/tykoas_controller.go

```diff
@@ -0,0 +1,324 @@
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	_, err = util.CreateOrUpdate(ctx, r.Client, tykOAS, func() error {
+		if !tykOAS.ObjectMeta.DeletionTimestamp.IsZero() {
+			markForDeletion = true
+			err := r.delete(ctx, tykOAS)
+			if err != nil {
+				return err
+			}
+
+			util.RemoveFinalizer(tykOAS, keys.TykOASFinalizerName)
+
+			return nil
+		}
+
+		util.AddFinalizer(tykOAS, keys.TykOASFinalizerName)
+
+		apiID, err = r.createOrUpdateTykOASAPI(ctx, tykOAS)
+		if err != nil {
+			r.Log.Error(err, "Failed to create/update Tyk OAS API", "name", client.ObjectKeyFromObject(tykOAS))
+			return err
+		}
+		return nil
+	})
+
+	if !markForDeletion {
+		var transactionInfo v1alpha1.TransactionInfo
+
+		if err == nil {
+			transactionInfo.Status = v1alpha1.Successful
+			transactionInfo.Time = metav1.Now()
+		} else {
+			reqA = queueAfter
+
+			transactionInfo.Status = v1alpha1.Failed
+			transactionInfo.Time = metav1.Now()
+			transactionInfo.Error = err.Error()
+		}
+
+		tykOAS.Status.LatestTransaction = transactionInfo
+
+		if err = r.updateStatus(ctx, tykOAS, apiID); err != nil {
+			r.Log.Error(err, "Failed to update status of Tyk OAS CRD", "name", client.ObjectKeyFromObject(tykOAS))
+			return ctrl.Result{RequeueAfter: reqA}, err
+		}
+	}
+
+	if err := klient.Universal.HotReload(ctx); err != nil {
+		r.Log.Error(err, "Failed to reload gateway", "name", client.ObjectKeyFromObject(tykOAS))
+		return ctrl.Result{RequeueAfter: reqA}, err
+	}
+
+	r.Log.Info("Completed reconciling Tyk OAS instance")
+
+	return ctrl.Result{RequeueAfter: reqA}, err
+}
+
+func (r *TykOASReconciler) createOrUpdateTykOASAPI(ctx context.Context, tykOASCrd *v1alpha1.TykOAS) (string, error) {
+	var cm v1.ConfigMap
+
+	ns := types.NamespacedName{
+		Name:      tykOASCrd.Spec.TykOAS.ConfigmapRef.Name,
+		Namespace: tykOASCrd.Spec.TykOAS.ConfigmapRef.Namespace,
+	}
+
+	err := r.Client.Get(ctx, ns, &cm)
+	if err != nil {
+		r.Log.Error(err, "Failed to fetch config map", "name", ns)
+		return "", err
+	}
+
+	tykOASDoc := cm.Data[tykOASCrd.Spec.TykOAS.ConfigmapRef.KeyName]
+
+	apiID, err := r.getAPIID(tykOASCrd, tykOASDoc)
+	if err != nil {
+		return "", err
+	}
+
+	if apiID == "" {
+		apiID = EncodeNS(client.ObjectKeyFromObject(tykOASCrd).String())
+	}
+
+	exists := klient.Universal.TykOAS().Exists(ctx, apiID)
+	if !exists {
+		if err = klient.Universal.TykOAS().Create(ctx, apiID, tykOASDoc); err != nil {
+			r.Log.Error(err, "Failed to create Tyk OAS API", "name", client.ObjectKeyFromObject(tykOASCrd))
+			return "", err
+		}
+	} else {
+		if err = klient.Universal.TykOAS().Update(ctx, apiID, tykOASDoc); err != nil {
+			r.Log.Error(err, "Failed to update Tyk OAS API", "name", client.ObjectKeyFromObject(tykOASCrd))
+			return "", err
+		}
+	}
+
+	return apiID, nil
+}
+
+func (r *TykOASReconciler) updateStatus(ctx context.Context, tykOASCrd *v1alpha1.TykOAS, apiID string) error {
+	var cm v1.ConfigMap
+
+	r.Log.Info("Updating status of Tyk OAS instance")
```

Comment:
From looking at this log, it is a bit unclear which OAS CR is being updated at the moment. Again, either we might consider using `log`, or indicating `tykOASCrd`'s metadata information in the log

Reply:
Resolved in [e3d178e](https://github.com/TykTechnologies/tyk-operator/pull/687/commits/e3d178ef016740ad1622181e3b43778641d34e37)